flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherHATest.java

```diff
@@ -225,13 +243,78 @@ public static JobGraph createNonEmptyJobGraph() {
 		return jobGraph;
 	}
 
-	private static class HATestingDispatcher extends TestingDispatcher {
+	@Test
+	public void testPersistedJobGraphWhenDispatcherIsShutDown() throws Exception {
+		final JobGraph jobGraph = createNonEmptyJobGraph();
+		final TestingHighAvailabilityServices highAvailabilityServices = new TestingHighAvailabilityServices();
 
-		@Nonnull
-		private final BlockingQueue<DispatcherId> fencingTokens;
+		final InMemorySubmittedJobGraphStore submittedJobGraphStore = new InMemorySubmittedJobGraphStore();
+		highAvailabilityServices.setSubmittedJobGraphStore(submittedJobGraphStore);
+
+		final TestingLeaderElectionService leaderElectionService = new TestingLeaderElectionService();
+		highAvailabilityServices.setDispatcherLeaderElectionService(leaderElectionService);
+
+		highAvailabilityServices.setJobMasterLeaderElectionService(jobGraph.getJobID(), new TestingLeaderElectionService());
+		highAvailabilityServices.setResourceManagerLeaderRetriever(new SettableLeaderRetrievalService());
+		highAvailabilityServices.setCheckpointRecoveryFactory(new StandaloneCheckpointRecoveryFactory());
 
-		HATestingDispatcher(RpcService rpcService, String endpointId, Configuration configuration, HighAvailabilityServices highAvailabilityServices, ResourceManagerGateway resourceManagerGateway, BlobServer blobServer, HeartbeatServices heartbeatServices, JobManagerMetricGroup jobManagerMetricGroup, @Nullable String metricQueryServicePath, ArchivedExecutionGraphStore archivedExecutionGraphStore, JobManagerRunnerFactory jobManagerRunnerFactory, FatalErrorHandler fatalErrorHandler, @Nonnull BlockingQueue<DispatcherId> fencingTokens) throws Exception {
-			super(rpcService, endpointId, configuration, highAvailabilityServices, resourceManagerGateway, blobServer, heartbeatServices, jobManagerMetricGroup, metricQueryServicePath, archivedExecutionGraphStore, jobManagerRunnerFactory, fatalErrorHandler);
+		final HATestingDispatcher dispatcher = createDispatcher(
+			highAvailabilityServices,
+			Dispatcher.DefaultJobManagerRunnerFactory.INSTANCE);
+
+		dispatcher.start();
+
+		try {
+			// grant leadership and submit a single job
+			final DispatcherId expectedDispatcherId = DispatcherId.generate();
+			leaderElectionService.isLeader(expectedDispatcherId.toUUID()).get();
+
+			final DispatcherGateway dispatcherGateway = dispatcher.getSelfGateway(DispatcherGateway.class);
+			final CompletableFuture<Acknowledge> submissionFuture = dispatcherGateway.submitJob(jobGraph, timeout);
+			submissionFuture.get();
+			assertThat(dispatcher.getNumberJobs(timeout).get(), is(1));
+
+			dispatcher.shutDown();
+			dispatcher.getTerminationFuture().get();
+
+			assertThat(submittedJobGraphStore.contains(jobGraph.getJobID()), is(true));
```

Comment:
Will it be deleted if it is not an HA setup?

Reply:
No it won't. But in the non-HA mode, we usually instantiate a `StandaloneSubmittedJobGraphStore` which is a no-op class.